### PR TITLE
Safeguard against sudden and dramatic cluster reconfigurations

### LINKED ISSUE #1185: Run probes to possibly detect container runtime
**What feature would you like to be added?**
Run probes to possibly detect container runtime and compare this information to the config. If docker is already running but conrainerd is configured, probably user is trying to blow up the cluster and kubeone should prevent this.

A follow up based on https://github.com/kubermatic/kubeone/pull/1180#issuecomment-739939128

### PR DESCRIPTION
**Pull request by kron4eg**
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1185 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Safeguard against sudden and dramatic cluster reconfigurations
```

### CODE REVIEW

File: pkg/scripts/os.go [hunk 1]
```diff
@@ -25,295 +25,6 @@ const (
...
-
-sudo systemctl daemon-reload
-sudo systemctl enable --now kubelet
-
-{{- if or .FORCE .KUBELET }}
-sudo systemctl restart kubelet
-{{- end }}
-`
-
 	kubeadmCoreOSTemplate = `
```
Comment (xmudrii) [inline pkg/scripts/os.go:28]: Can you also move CoreOS/Flatcar templates to a dedicated file for consistency?

**Comment by kron4eg** [comment]
/hold

I figured, the safeguard that iterates over all nodes is too strict since there also might be machine-controller controlled nodes that the container runtimes will be different

File: pkg/state/cluster.go [hunk 2]
```diff
@@ -242,7 +244,17 @@ func (h *Host) WorkerHealthy() bool {
 }
 
 func (h *Host) healthy() bool {
-	return h.IsInCluster && h.ContainerRuntime.Healthy() && h.Kubelet.Healthy()
+	var crStatus bool
+
+	if h.ContainerRuntimeContainerd.IsProvisioned() {
```
Comment (xmudrii) [inline pkg/state/cluster.go:249]: I don't understand this check. How do we know that Docker is installed and supposed to be healthy if we just check is containerd provisioned? Shouldn't we check here is Docker provisioned instead?
```suggestion
	if h.ContainerRuntimeDocker.IsProvisioned() {
```

Comment (kron4eg) [reply pkg/state/cluster.go:249]: Yep, you're absolutely right, looks like I've messed up when was moving the code around. Nice catch!

File: pkg/scripts/os.go (hunk 1 above)
Comment (kron4eg) [reply pkg/scripts/os.go:28]: I will actually move this extraction to the dedicated PR, as this move is not strictly related to the safeguard :)

File: pkg/apis/kubeone/helpers.go [hunk 3]
```diff
@@ -52,6 +52,16 @@ func (c KubeOneCluster) Followers() []HostConfig {
 	return followers
 }
 
+func (c *KubeOneCluster) OwnNode(name string) bool {
```
Comment (xmudrii) [inline pkg/apis/kubeone/helpers.go:55]: `OwnNode` is a little bit confusing name to me. I would propose that we rename it to something like `NodeExists`. Also, please add a comment describing what this function does.

File: pkg/tasks/probes.go [hunk 4]
```diff
@@ -37,20 +37,44 @@ import (
...
+	systemdShowExecStartCMD = `systemctl show %s -p ExecStart`
 
-	dockerVersionDPKG   = `dpkg-query --show --showformat='${Version}' docker-ce | cut -d: -f2 | cut -d~ -f1`
-	dockerVersionRPM    = `rpm -qa --queryformat '%{RPMTAG_VERSION}' docker-ce`
-	dockerVersionAmazon = `rpm -qa --queryformat '%{RPMTAG_VERSION}' docker`
+	kubeletInitializedCMD = `test -f /etc/kubernetes/kubelet.conf`
+)
 
-	containerdVersion = `containerd --version | awk '{print $3}'`
+func safeguard(s *state.State) error {
```
Comment (xmudrii) [inline pkg/tasks/probes.go:46]: Should we allow users to skip this safeguard with the `--force` flag?

Comment (kron4eg) [reply pkg/tasks/probes.go:46]: Not now, but maybe in the future when possibly we will figure out the way to migrate between runtimes.

**Review by xmudrii — APPROVED** [review]
/lgtm

**Pull request merged**